crates/bevy_utils/src/parallel_queue.rs

```diff
@@ -0,0 +1,92 @@
+use core::{
+    cell::Cell,
+    ops::{Deref, DerefMut, Drop},
+};
+use thread_local::ThreadLocal;
+
+/// A cohesive set of thread-local values of a given type.
+///
+/// Mutable references can be fetched if `T: Default` via [`Parallel::get`].
+#[derive(Default)]
+pub struct Parallel<T: Send> {
+    locals: ThreadLocal<Cell<T>>,
+}
+
+impl<T: Send> Parallel<T> {
+    /// Gets a mutable iterator over all of the per-thread queues.
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = &'_ mut T> {
+        self.locals.iter_mut().map(|cell| cell.get_mut())
+    }
+
+    /// Clears all of the stored thread local values.
+    pub fn clear(&mut self) {
+        self.locals.clear();
+    }
+}
+
+impl<T: Default + Send> Parallel<T> {
+    /// Takes the thread-local value and replaces it with the default.
+    #[inline]
+    pub fn get(&self) -> ParRef<'_, T> {
```

Comment:
could we make this take &mut self to avoid the footgun?

Reply:
This would make it impossible to send this to send `&Parallel<T>` to multiple threads.

Reply:
ah right.

Overall I'm not that worried about the footgun. All the internal uses have been fairly localized. The footgun is only really really problematic if someone is doing something like calling `get` from multiple functions. Otherwise you'll be able to detect the problem fairly easily.